From f989d311702a5f22e54bfefb740c75ae4eadb0ad Mon Sep 17 00:00:00 2001
From: Dan Nica <dan.nica@nxp.com>
Date: Thu, 21 Jul 2022 07:52:22 +0300
Subject: [PATCH 11/42] s32: mmc: Touch only the relevant INT_STATUS(_EN) bits

Only enable and clear the status bits that are actually used. Mainly, this
avoids writting reserved fields in INT_STATUS when clearing the flags.

Issue: ALB-9078
Upstream-Status: Pending 

Signed-off-by: Dan Nica <dan.nica@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/nxp/s32/mmc/s32_mmc.c | 15 ++++++++-------
 1 file changed, 8 insertions(+), 7 deletions(-)

diff --git a/drivers/nxp/s32/mmc/s32_mmc.c b/drivers/nxp/s32/mmc/s32_mmc.c
index 185300677..5b41f0c72 100644
--- a/drivers/nxp/s32/mmc/s32_mmc.c
+++ b/drivers/nxp/s32/mmc/s32_mmc.c
@@ -57,7 +57,6 @@
 #define SYS_CTRL_DVS_MASK		(SYS_CTRL_DVS(0xf))
 
 #define USDHC_INT_STATUS		(USDHC + 0x30)
-#define INT_STATUS_CLEAR_ALL		(0xffffffff)
 #define INT_STATUS_DMAE			BIT(28)
 #define INT_STATUS_DEBE			BIT(22)
 #define INT_STATUS_DCE			BIT(21)
@@ -72,9 +71,13 @@
 					 INT_STATUS_CCE | INT_STATUS_CTOE)
 #define INT_STATUS_DATA_ERROR		(INT_STATUS_DMAE | INT_STATUS_DEBE | \
 					 INT_STATUS_DCE | INT_STATUS_DTOE)
+#define INT_STATUS_CLEARMASK		(INT_STATUS_CC | INT_STATUS_TC | \
+					 INT_STATUS_CMD_ERROR | \
+					 INT_STATUS_DATA_ERROR)
+
 #define USDHC_INT_STATUS_EN		(USDHC + 0x34)
-#define INT_STATUS_EN_BRRSEN		BIT(5)
-#define INT_STATUS_EN_BWRSEN		BIT(4)
+#define INT_STATUS_EN_ENABLEMASK	INT_STATUS_CLEARMASK
+
 #define USDHC_INT_SIGNAL_EN		(USDHC + 0x38)
 
 #define USDHC_MIX_CTRL			(USDHC + 0x48)
@@ -178,9 +181,7 @@ static void s32_mmc_init(void)
 
 	s32_mmc_set_clk(IDENTIFICATION_MODE_FREQUENCY);
 
-	regdata = 0xffffffff & ~(INT_STATUS_EN_BRRSEN | INT_STATUS_EN_BWRSEN);
-	mmio_write_32(USDHC_INT_STATUS_EN, regdata);
-
+	mmio_write_32(USDHC_INT_STATUS_EN, INT_STATUS_EN_ENABLEMASK);
 	mmio_write_32(USDHC_PROT_CTRL, PROT_CTRL_EMODE_LE);
 
 	regdata = mmio_read_32(USDHC_SYS_CTRL);
@@ -198,7 +199,7 @@ static int s32_mmc_send_cmd(struct mmc_cmd *cmd)
 	uint32_t regdata;
 	uint64_t adtc_mask;
 
-	mmio_write_32(USDHC_INT_STATUS, INT_STATUS_CLEAR_ALL);
+	mmio_write_32(USDHC_INT_STATUS, INT_STATUS_CLEARMASK);
 	while (mmio_read_32(USDHC_PRES_STATE) &
 	       (PRES_STATE_CDIHB | PRES_STATE_CIHB | PRES_STATE_DLA))
 		;
-- 
2.17.1

